### ISSUE TITLE
Potential macOS version discussion

### ISSUE THREAD

**Issue by petmac** [decaf-emu/decaf-emu#426]
Hi,

I'm interested in doing some or all of the work required to port decaf-emu to macOS.

I've done some work on it independently, and the emulator builds and runs.

The current roadblock is that macOS only supports OpenGL 4.1 rather than the current target of 4.5. This doesn't stop the emulator running (after some version number changes) but it does prevent anything from being drawn. Some functions and some GLSL features are missing.

In order to not waste anyone's time, I wanted to discuss the following scenarios:

1. A macOS version is not wanted, in which case I'll just stop working on it.
2. A macOS version is wanted, but it's not acceptable to downgrade the OpenGL requirement on Windows or Linux to 4.1. Assuming macOS can support everything decaf-emu needs, a separate macOS-compatible renderer (i.e., Metal or OpenGL 4.1) would be required.
3. A macOS version is wanted, and it's acceptable to downgrade to OpenGL 4.1.

Thanks,
Peter

**Comment by brett19** [comment]
Hey @petmac,

@exjam and I have had this discussion on numerous occasions, as we both actually use MBP's for work.  The main problem with downgrading OpenGL to 4.1 is that it legitimately does not support the features that are required to render the WiiU games, it also has some performance implications.

We had previously considered implementing a Metal rendering engine, similar to how we've implemented our DX12 rendering engine, but the interop between Obj-C and C++11 is non-trivial (although just searching now, I found this new project which looks promising: https://github.com/naleksiev/mtlpp).  I have however read a few things as well about Obj-C <-> C++ interop being difficult to get leak free due to differing semantics.

All of this being said, we would be ecstatic to see a community member contribute a Metal rendering backend.  I'd be more than happy to provide any necessary support towards the goal of implementing a Metal backend.

Cheers, Brett

**Comment by exjam** [comment]
Yeah it basically comes to Apple really dropping the ball on it's graphics driver situation. Even Metal isn't sufficient as it lacks features such as Geometry Shaders. And looking at https://en.wikipedia.org/wiki/OpenGL#OpenGL_4.1 looks like there has been some useful OpenGL stuff between 4.1 and 4.5.

I mean you can try - but it's going to be rather challenging to emulate the missing features.

**Comment by Gamedev1909** [comment]
Well use metal to vulkan wrapper its free called moltelvk works just like vulkan

**Comment by exjam** [comment]
MoltenVK does not support geometry shaders.

**Comment by Gamedev1909** [comment]
Well dam oh well new metal build is coming with 10.13 as it will be a major build thank god

**Comment by petmac** [comment]
Thanks, guys.

I've not looked into it in great detail, but it's possible that geometry shaders could be translated to compute shaders. 

Apple added tessellation with macOS Sierra, so as @gamedev1909 said, maybe it'll come in a future update. 

Do many games use geometry shaders? Maybe it's something we could live with not supporting on Mac.

What I'd like to do then is fix up my code so it works with the latest master, add a macOS build to Travis, disable OpenGL for Mac, stub out an empty rendering engine, then submit a pull request with that. That way, Mac support can come in incrementally after that. Would that be OK?

**Comment by exjam** [comment]
Yeah not all games use geometry shaders, of my collection about 50% of them use GX2SetGeometryShader. Including some of the more popular titles such as the new Zelda, MK8 and Smash.

It is possible to emulate geometry shaders using compute, I know someone trying to do that exact thing for a different project, they have it working but at pretty poor performance for now.

I would not object to a partially functional backend for OSX though - we can warn users that games using X features are not supported.

**Comment by brett19** [comment]
@petmac said:
I got my code rebased on top of the libgpu changes. Thanks for letting
me know before I got too far along.

The Metal rendering back end and debugger UI renderer will probably
both need to be passed a shared command queue (an id<MTLCommandQueue>)
on construction/initialisation and a texture to render to on draw.

I've worked around this somewhat, but it clearly breaks the existing
debugger UI interface. That is, I won't be using either of the
existing initialise/draw methods, and will instead call different ones
from decafsdl_metal.mm.

I hope this makes sense.

**Comment by brett19** [comment]
If the abstraction that exists at the moment is insufficient to enable implementation of a Metal renderer, then it is likely that the abstraction should be revised to provide a more abstracted API so that it fits all of the APIs equally.  Note that initialize and draw were intentionally designed to be abstracted enough to support any number of backends.  You may want to inspect the file I've linked below and build a similar file/class for Metal.

https://github.com/decaf-emu/decaf-emu/blob/master/src/libdecaf/src/debugger/debugger_ui_opengl.cpp

**Comment by petmac** [comment]
Thanks @brett19 

As you suggest, I'm building a similar file to debugger_ui_opengl.cpp but for Metal.

The OpenGL abstraction is somewhat leaky because it relies on the existence of a globally accessible current OpenGL context. Like D3D, Metal doesn't have such a thing, but it does have an MTLDevice which you can pass around, or you can pass around a MTLCommandQueue, and if you need the device you can get it from that.

**Comment by brett19** [comment]
Oh.... Looks like most of my dx12 work is still sitting over on its own branch.  Here is a vague idea of how I did it for DX12 though:
https://github.com/decaf-emu/decaf-emu/blob/dx12/src/libdecaf/src/debugger/debugger_ui_dx12.cpp

**Comment by exjam** [comment]
Meh, first step is always to get something working for now and once we fully understand the problem then we can make a clean interface which discriminates against no graphics backend. An equal opportunities interface if you will 🏳️‍🌈

**Comment by petmac** [comment]
I wanted to post a quick update.

This is still being worked on, on and off. What's done:

- Debug UI rendering
- Scan buffers, swapping, blitting them to the SDL window
- Some basic render pass and pipeline management

And things that can be done next, off the top of my head:

- Color buffer conversion
- Register translation to render state
- Shader translation to MSL
- Vertex data copying/conversion

**Comment by Gamedev1909** [comment]
Have you looked into metal 2 api it had a ton of api features added like geo shaders

**Comment by petmac** [comment]
I have a bit, as much as was mentioned on WWDC 2017 videos anyway. Some nice features, but my machine is a MacBook Air so I may not be able to take advantage of them all. Hopefully there will be something of use though.

**Comment by Gamedev1909** [comment]
I will do any metal 2 testing you need with my hackintosh my 290x is fully metal compatible

**Comment by rien333** [comment]
How far along is this? Can the current source deliver anything playable at all? Or is the OpenGL version to limiting? Any updates on the potential use of Metal?

**Comment by Gamedev1909** [comment]
Update moltenvk is going open source

**Comment by VelocityRa** [comment]
...source?

**Comment by Gamedev1909** [comment]
That image is the main dev if the image showed up

**Comment by Gamedev1909** [comment]
https://imgur.com/a/8K24g Here i see you cant send images in github emails

**Comment by exjam** [comment]
The author is now working on his own PS Vita emulator, so I presume this port is dead.

https://github.com/Vita3K/Vita3K

**Comment by petmac** [comment]
Dead is maybe a bit harsh, but I'm afraid the Vita emulator started coming together, and with family commitments squeezing what was left of my time I focused my energy on that.

I'd like to come back to it, but can't say when.

I found it quite hard to keep the macOS version in sync with other renderer changes, but if you guys are working on other renderers it should become easier in the future.

Or maybe MoltenVK will be best, I'm not sure.
